### PR TITLE
Created Orchestrator component for applicationInstances

### LINKED ISSUE #12: Need to write Orchestrator's component for ApplicationInstances.
For example it can be ApplicationInstancesOrchestrator.cs inside Orchestrators folder

- It should scan section ApplicationInstances in dg-cluster.config.json, retrieve it contents and ensure instances in memory via reflection Activator.CreateInstance. 
- it would be great if it has internal dictionary of application instances (need to think about dictionary key - as option we can use Name + Type currently)

### PR DESCRIPTION
**Pull request by KorolS**
#12 Added Scanners for config file (ConfigJSONFileScanner can scan any json file into nested dictionary, this solution can be extended in future to scan json via deserialization), scanner for instances, scanner for Assembly.
Created ApplicationInstancesOrchestrator to ensure instances in memory.

### CODE REVIEW

File: source/DG.Core/Scanners/AssemblyScanner.cs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+        public IEnumerable<Type> GetAssemblyTypes(string assemblyPath)
+        {
+            Assembly assembly = Assembly.LoadFrom(assemblyPath);
+            return assembly.GetTypes();
+        }
+
+        public Type GetType(string assemblyPath, string namespaceName, string typeName)
+        {
+            Assembly assembly = Assembly.LoadFrom(assemblyPath);
+            return assembly.GetType(namespaceName + "." + typeName);
```
Comment (motpfofs) [inline source/DG.Core/Scanners/AssemblyScanner.cs]: We can create new variable 
`private char nameAndTypeSeparator = '.';` 
and use it here 
`assembly.GetType(namespaceName + nameAndTypeSeparator + typeName)`

File: source/DG.Core/Scanners/ApplicationInstancesScanner.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+                {
+                    foreach (var field_values in instanceFields)
+                    {
+                        IDictionary<string, object> instanceFieldsData = new Dictionary<string, object>();
+                        foreach (var field_value in (IDictionary<string, object>)field_values)
+                        {
+                            instanceFieldsData.Add(field_value.Key, field_value.Value);
+                        }
+
+                        var instanceType = instanceFieldsData["Type"];
```
Comment (motpfofs) [inline source/DG.Core/Scanners/ApplicationInstancesScanner.cs]: We can create variable
`private string instanceTypeFieldName = "Type"`
and use it here
`var instanceType = instanceFieldsData[instanceTypeFieldName];`
and later:
`return this.GetFieldValuesFromApps(instanceTypeFieldName);`
Same approach can be applied to `"Name"`

File: source/DG.Core/Scanners/ConfigJSONFileScanner.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        public ConfigJSONFileScanner(string pathToConfig)
+        {
+            this.configPath = pathToConfig;
+        }
+
+        public IDictionary<string, object> GetKeyValuesFromAllSections()
+        {
+            var configurationSections = this.GetBaseConfigSections();
+            IDictionary<string, object> sectionValues = new Dictionary<string, object>();
+            foreach (ConfigurationSection configurationSection in configurationSections)
```
Comment (motpfofs) [inline source/DG.Core/Scanners/ConfigJSONFileScanner.cs:21]: We can increase readability by using `var`:
`var sectionValues = new Dictionary<string, object>();`
`foreach (var configurationSection in configurationSections)`

Comment (chesdenis) [reply source/DG.Core/Scanners/ConfigJSONFileScanner.cs:21]: I guess we can try to use IOptions<T> here https://blog.zwezdin.com/2016/configuring-aspnet-core-apps-with-ioptions/

File: source/DG.Core/Scanners/AssemblyScanner.cs (hunk 1 above)
Comment (chesdenis) [reply source/DG.Core/Scanners/AssemblyScanner.cs]: what if we use extensions here?

**Review by chesdenis — CHANGES_REQUESTED** [review]
There is no any tests in PR: 
Please add unit tests for orchestrator to cover this features for example:
ShouldCreateApplicationInstance - expect to see instance in memory (need to look into Orchestrator internal state somehow)
ShouldThrowOnceCantFindType - occurs when orchestrator can't find type in assembly.

File: source/DG.Core/Scanners/AssemblyScanner.cs (hunk 1 above)
Comment (KorolS) [reply source/DG.Core/Scanners/AssemblyScanner.cs]: I have reworked this part

File: source/DG.Core/Scanners/ApplicationInstancesScanner.cs (hunk 2 above)
Comment (KorolS) [reply source/DG.Core/Scanners/ApplicationInstancesScanner.cs]: I have reworked this part

File: source/DG.Core/Scanners/ConfigJSONFileScanner.cs (hunk 3 above)
Comment (KorolS) [reply source/DG.Core/Scanners/ConfigJSONFileScanner.cs:21]: I haven't find good decision to use IOptions in DG.Core project component.

**Comment by KorolS** [comment]
Added tests for orchestrator

**Review by chesdenis — APPROVED** [review]

**Pull request merged**